File: packages/ui-compose/ui/src/test/kotlin/com/kinetixui/ui/blocks/TableToolbarBlockTest.kt
package com.kinetixui.ui.blocks

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.test.junit4.v2.createComposeRule
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.unit.dp
import com.kinetixui.ui.KinetixButton
import com.kinetixui.ui.KinetixDropdownMenu
import com.kinetixui.ui.KinetixInput
import com.kinetixui.ui.KinetixSelectItem
import com.kinetixui.ui.KinetixSelectTrigger
import com.kinetixui.ui.KinetixTable
import com.kinetixui.ui.KinetixTableBody
import com.kinetixui.ui.KinetixTableCell
import com.kinetixui.ui.KinetixTableHead
import com.kinetixui.ui.KinetixTableHeader
import com.kinetixui.ui.KinetixTableRow
import com.kinetixui.ui.KinetixTheme
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

// The "Table + toolbar" block — see TestimonialBlockTest.kt for how block fixtures work.
//
// The filter is a KinetixSelectTrigger opening a KinetixDropdownMenu, which is how a select is composed on
// Compose — there is no single KinetixSelect composable, because the platform's own anchored-menu mechanism
// already provides one. The status cell is plain text rather than the web version's Badge: KinetixTableCell
// takes a String, not a content slot.
//
// kx-block:start
data class Invoice(val id: String, val status: String, val amount: String)

@Composable
fun TableToolbarBlock(
    invoices: List<Invoice> = listOf(
        Invoice("INV-001", "Paid", "$250.00"),
        Invoice("INV-002", "Pending", "$150.00"),
        Invoice("INV-003", "Paid", "$350.00"),
    ),
    onAdd: () -> Unit = {},
) {
    var query by remember { mutableStateOf("") }
    var status by remember { mutableStateOf("All statuses") }
    var open by remember { mutableStateOf(false) }
    val statuses = listOf("All statuses", "Paid", "Pending")

    Column {
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            KinetixInput(query, { query = it }, Modifier.weight(1f), placeholder = "Search invoices…")
            KinetixDropdownMenu(
                visible = open,
                onDismissRequest = { open = false },
                anchor = { KinetixSelectTrigger(status, { open = true }, Modifier.width(144.dp)) },
            ) {
                statuses.forEach { option ->
                    KinetixSelectItem(option, option == status, { status = option; open = false })
                }
            }
            KinetixButton(onAdd) { Text("Add invoice") }
        }
        KinetixTable(Modifier.padding(top = 16.dp)) {
            KinetixTableHeader {
                KinetixTableRow {
                    KinetixTableHead("Invoice")
                    KinetixTableHead("Status")
                    KinetixTableHead("Amount")
                }
            }
            KinetixTableBody {
                invoices.forEach { invoice ->
                    KinetixTableRow {
                        KinetixTableCell(invoice.id)
                        KinetixTableCell(invoice.status)
                        KinetixTableCell(invoice.amount)
                    }
                }
            }
        }
    }
}
// kx-block:end

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class TableToolbarBlockTest {
    @get:Rule
    val rule = createComposeRule()

    @Test
    fun renders_the_toolbar_and_every_row() {
        rule.setContent { KinetixTheme(darkTheme = false) { TableToolbarBlock() } }
        rule.onNodeWithText("Search invoices…").assertExists()
        rule.onNodeWithText("Add invoice").assertExists()
        rule.onNodeWithText("Invoice").assertExists()
        rule.onNodeWithText("INV-002").assertExists()
        rule.onNodeWithText("$350.00").assertExists()
    }
}

File: packages/ui-flutter/test/blocks/table_toolbar_block_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:kinetix_ui/kinetix_ui.dart';

// The "Table + toolbar" block — see testimonial_block_test.dart for how block fixtures work.
//
// kx-block:start
class Invoice {
  const Invoice(this.id, this.status, this.amount, {this.paid = false});

  final String id;
  final String status;
  final String amount;
  final bool paid;
}

class TableToolbarBlock extends StatefulWidget {
  const TableToolbarBlock({super.key, this.invoices = defaultInvoices, this.onAdd});

  static const List<Invoice> defaultInvoices = <Invoice>[
    Invoice('INV-001', 'Paid', r'$250.00', paid: true),
    Invoice('INV-002', 'Pending', r'$150.00'),
    Invoice('INV-003', 'Paid', r'$350.00', paid: true),
  ];

  final List<Invoice> invoices;
  final VoidCallback? onAdd;

  @override
  State<TableToolbarBlock> createState() => _TableToolbarBlockState();
}

class _TableToolbarBlockState extends State<TableToolbarBlock> {
  final TextEditingController _query = TextEditingController();
  String _status = 'all';

  @override
  void dispose() {
    _query.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        Row(
          children: <Widget>[
            Expanded(child: KinetixInput(controller: _query, placeholder: 'Search invoices…')),
            const SizedBox(width: 8),
            SizedBox(
              width: 144,
              child: KinetixSelect<String>(
                value: _status,
                options: const <KinetixSelectOption<String>>[
                  KinetixSelectOption<String>('all', 'All statuses'),
                  KinetixSelectOption<String>('paid', 'Paid'),
                  KinetixSelectOption<String>('pending', 'Pending'),
                ],
                onChanged: (String? v) => setState(() => _status = v ?? 'all'),
              ),
            ),
            const SizedBox(width: 8),
            KinetixButton(onPressed: widget.onAdd, child: const Text('Add invoice')),
          ],
        ),
        const SizedBox(height: 16),
        KinetixTable(
          children: <Widget>[
            const KinetixTableRow(
              isHeader: true,
              cells: <Widget>[
                KinetixTableHead('Invoice'),
                KinetixTableHead('Status'),
                KinetixTableHead('Amount'),
              ],
            ),
            for (final Invoice invoice in widget.invoices)
              KinetixTableRow(
                cells: <Widget>[
                  KinetixTableCell(child: Text(invoice.id)),
                  KinetixTableCell(
                    child: KinetixBadge(
                      invoice.status,
                      variant: invoice.paid ? KinetixBadgeVariant.subtle : KinetixBadgeVariant.outline,
                    ),
                  ),
                  KinetixTableCell(child: Text(invoice.amount)),
                ],
              ),
          ],
        ),
      ],
    );
  }
}
// kx-block:end

void main() {
  testWidgets('renders the toolbar and every row', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: KinetixTheme(
          brightness: Brightness.light,
          child: const Scaffold(body: SingleChildScrollView(child: TableToolbarBlock())),
        ),
      ),
    );

    expect(find.text('Add invoice'), findsOneWidget);
    expect(find.text('Invoice'), findsOneWidget);
    expect(find.text('INV-002'), findsOneWidget);
    expect(find.text(r'$350.00'), findsOneWidget);
  });
}
